src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
-            if (_trackUnassignments)
+            TLocalState result;
+            if (_labels.TryGetValue(label, out result))
             {
-                Optional<TLocalState> oldTryState = _tryState;
-                _tryState = AllBitsSet();
-                VisitFinallyBlock(finallyBlock, ref unsetInFinally);
-                var tempTryStateValue = _tryState.Value;
-                IntersectWith(ref unsetInFinally, ref tempTryStateValue);
-                if (oldTryState.HasValue)
-                {
-                    var oldTryStateValue = oldTryState.Value;
-                    IntersectWith(ref oldTryStateValue, ref tempTryStateValue);
-                    oldTryState = oldTryStateValue;
-                }
+                return result;
+            }
 
-                _tryState = oldTryState;
+            result = UnreachableState();
+            _labels.Add(label, result);
+            return result;
+        }
+
+        /// <summary>
+        /// Return to the caller the set of pending return statements.
+        /// </summary>
+        /// <returns></returns>
+        protected virtual ImmutableArray<PendingBranch> RemoveReturns()
+        {
+            ImmutableArray<PendingBranch> result;
+            if (_trackExceptions)
+            {
+                // when we are tracking exceptions, we use pendingBranches[0] to
+                // track exception states.
+                result = PendingBranches
+                        .Where(b => b.Branch != null)
+                        .AsImmutableOrNull();
+
+                var oldExceptions = PendingBranches[0];
+                Debug.Assert(oldExceptions.Branch == null);
+                PendingBranches.Clear();
+                PendingBranches.Add(oldExceptions);
             }
             else
             {
-                VisitFinallyBlock(finallyBlock, ref unsetInFinally);
+                result = PendingBranches.ToImmutable();
+                PendingBranches.Clear();
             }
+
+            // The caller should have handled and cleared labelsSeen.
+            Debug.Assert(_labelsSeen.Count == 0);
+            return result;
         }
 
-        protected virtual void VisitTryBlock(BoundStatement tryBlock, BoundTryStatement node, ref TLocalState tryState)
+        /// <summary>
+        /// Set the current state to one that indicates that it is unreachable.
+        /// </summary>
+        protected void SetUnreachable()
         {
-            VisitStatement(tryBlock);
+            this.State = UnreachableState();
         }
 
-        protected virtual void VisitCatchBlock(BoundCatchBlock catchBlock, ref TLocalState finallyState)
+        protected void VisitLvalue(BoundExpression node)
         {
-            if (catchBlock.ExceptionSourceOpt != null)
+            if (_trackRegions && node == this.firstInRegion && this.regionPlace == RegionPlace.Before) EnterRegion();
+            switch (node?.Kind)
```

Comment:
1. hard to read hte 'if' esp as the code it conditions is so far to the side.  
2. needs comment explaining why this region check is here.  the previous location i saw the region check made sense to me.  i'm not sure why we need this here.

Reply:
3. is this code identical to previous region handling code?  if so, can we share?